File: src/top150/n21_merge_two_sorted_list/Solution.go
package n21_merge_two_sorted_list

func main() {

}

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	if list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			list1.Next = mergeTwoLists(list1.Next, list2)
			return list1
		} else {
			list2.Next = mergeTwoLists(list1, list2.Next)
			return list2
		}
	}
	if list1 == nil {
		return list2
	} else {
		return list1
	}
}

File: src/top150/n21_merge_two_sorted_list/Solution.java
package top150.n21_merge_two_sorted_list;

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) {
        this.val = val;
    }
    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

public class Solution {

    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 != null && list2 != null) {
            if (list1.val < list2.val) {
                list1.next = mergeTwoLists(list1.next, list2);
                return list1;
            } else {
                list2.next = mergeTwoLists(list1, list2.next);
                return list2;
            }
        }

        return list1 == null ? list2 : list1;
    }

    public static void main(String[] args) {
        var s = new Solution();
        var r = s.mergeTwoLists(
                new ListNode(1, new ListNode(2, new ListNode(4))),
                new ListNode(1, new ListNode(3, new ListNode(4)))
                );

        while (r != null) {
            System.out.println(r.val);
            r = r.next;
        }
    }
}
